### ISSUE TITLE
ONNX graph exporter issues from TensorRT 5 perspective

### ISSUE THREAD

**Issue by drnikolaev** [pytorch/pytorch#16908]
## 🐛 Bug

ONNX model generated by torch-jit-export doesn't seem to be as good as it might be. Let's compare the "native" one with the one generated by PyTorch (this is RN50):

![caffe2_reshape](https://user-images.githubusercontent.com/7504104/52510076-7729a580-2baf-11e9-8329-09df8154a793.png)

![screenshot from 2019-02-08 14-44-13](https://user-images.githubusercontent.com/7504104/52510231-0df66200-2bb0-11e9-9939-7710ef1491fc.png)

Here is the parser error: 
```
While parsing node number 176 [Gather]:
4
ERROR: onnx2trt_utils.hpp:276 In function convert_axis:
[8] Assertion failed: axis >= 0 && axis < nbDims
[E] failed to parse onnx file
[E] Engine could not be created
[E] Engine could not be created
```

This is because in general we do not allow Unsqueeze and Gather operations to be done across axis 0, because in ONNX that is the batch dimension and TRT does not handle operations across the batch dimension.

In addition, the second Unsqueeze has a constant value of 18446744073709551615, which is incorrect.

Finally in TRT for reshaping we expect the new shape to be a weights object (i.e. a constant) instead of a tensor object (i.e. an output from a previous layer). This is because we do not know the values of the tensor until runtime which means we cannot pass the new shape into the reshape layer with a tensor. This is what the pytorch generated ONNX graph is doing currently doing.

At the first picture we see a resnet50 model that can be parsed and run by TRT into completion. The reshaped shape is passed as a constant, which means we bypass all the of the issues stated above. Is it possible to have the pytorch exporter do the same?

**Comment by drnikolaev** [comment]
Just a quick summary:


- For the reshape node instead of passing in the new shape as a constant, exporter should build the shape through a subgraph commonly seen with the flow **output -> shape -> gather/slice -> unsqueezes [with buggy constants] -> concat -> reshape**
- In addition it seems like Constant exporter has a bug where the value of the constant is not shown in the graph or is incorrectly exported into the graph.

How (TRT would like) this to be fixed:

Exporter should know the expected new shape, have it as a constant that's passed into the reshape layer.  New flow should look like **output -> reshape (with new shape information known as a constant)** This will also increase performance as we're skipping all the unnecessary layers.

**Comment by soumith** [comment]
cc: @houseroad @zrphercule

**Comment by gottbrath** [comment]
@houseroad @zrphercule  -- any feedback on this analysis from the TRT team? Do we agree that the unsqueeze constant is wrong and that we can represent the model using the approach described above?

**Comment by gottbrath** [comment]
possibly also related to https://github.com/pytorch/pytorch/issues/17606 ?

**Comment by houseroad** [comment]
@gottbrath seems unrelated. This is an interesting problem. I guess this is the culprit: https://github.com/pytorch/vision/blob/master/torchvision/models/resnet.py#L163.

I think using flatten should solve the problem.

**Comment by yinxiaochuan** [comment]
Same problem. I think the reason of this issue is that the exporter cannot calculate the shape of tensor dynamically.   

I try to convert the alxnet onnx model following the tutorial steps in 
https://github.com/onnx/tutorials/blob/master/tutorials/PytorchOnnxExport.ipynb

The same issue happens when I convert the onnx model to the tensorrt model using:
''' 
onnx2trt my_model.onnx -o my_engine.trt
'''
After I specify the number of shape of x by replace the https://github.com/pytorch/vision/blob/efa4322da54b0424a68b5d6250411823fcee1952/torchvision/models/alexnet.py#L46 with  x = x.view(1, 256 * 6 * 6) , the model can be convert successfully.

**Comment by gottbrath** [comment]
Discussion with some folks (nvidia, fb, ms) -- we think that the large integer value for unsqueeze is separate from the other parts of this bug and it would be clearer to spin that part off into a separate issue (which may be something we can work on quickly and close).

**Comment by borisfom** [comment]
The culprit is at line 182:
 x = x.view(x.size(0), -1)
If there is another operation that does the same but does not refer to size(0) explicitly, that would be a workaround. 

> @gottbrath seems unrelated. This is an interesting problem. I guess this is the culprit: https://github.com/pytorch/vision/blob/master/torchvision/models/resnet.py#L163.
> I think using flatten should solve the problem.

**Comment by spandantiwari** [comment]
> Discussion with some folks (nvidia, fb, ms) -- we think that the large integer value for unsqueeze is separate from the other parts of this bug and it would be clearer to spin that part off into a separate issue (which may be something we can work on quickly and close).

I have investigated the particular issue of large integer value (18446744073709551615) as input to `Unsqueeze`. The large integer shows up not only in the context of `Unsqueeze` or `Reshape`, but anytime there a negative value stored in an `onnx::Constant` of tensor type `int64`. However, this particular scenario is not a bug in PyTorch ONNX exporter, but seems to be an issue in the ONNX graph visualization tool Netron. @drnikolaev - I am guessing the snapshot attached above is from Netron? This issue happens in Netron when there's a negative value stored as `int64` tensor in an `onnx::Constant`.  The value is serialized to the ONNX protobuf correctly. But Netron seems to be misinterpreting `int64` as `uint64` and displaying the wrong value in the graph. The type (`int64`) is displayed correctly. 

I have opened an issue for Netron. 
https://github.com/lutzroeder/netron/issues/264

For this scenario, the ONNX graph exported by PyTorch exporter seems to be correct.
